doc/man3/OSSL_SERIALIZER_CTX_new_by_EVP_PKEY.pod

```diff
@@ -0,0 +1,128 @@
+=pod
+
+=head1 NAME
+
+OSSL_SERIALIZER_CTX_new_by_EVP_PKEY,
+OSSL_SERIALIZER_CTX_set_cipher_name,
+OSSL_SERIALIZER_CTX_set_passphrase,
+OSSL_SERIALIZER_CTX_set_passphrase_cb,
+OSSL_SERIALIZER_CTX_set_passphrase_ui,
+OSSL_SERIALIZER_PUBKEY_TO_PEM_PQ,
+OSSL_SERIALIZER_PrivateKey_TO_PEM_PQ,
+OSSL_SERIALIZER_Parameters_TO_PEM_PQ,
+OSSL_SERIALIZER_PUBKEY_TO_TEXT_PQ,
+OSSL_SERIALIZER_PrivateKey_TO_TEXT_PQ,
+OSSL_SERIALIZER_Parameters_TO_TEXT_PQ
+- Serializer routines to serialize EVP_PKEYs
+
+=head1 SYNOPSIS
+
+ #include <openssl/serializer.h>
+
+ OSSL_SERIALIZER_CTX *OSSL_SERIALIZER_CTX_new_by_EVP_PKEY(const EVP_PKEY *pkey,
+                                                          const char *propquery);
+
+ int OSSL_SERIALIZER_CTX_set_cipher_name(OSSL_SERIALIZER_CTX *ctx,
+                                         const char *cipher_name);
+ int OSSL_SERIALIZER_CTX_set_passphrase(OSSL_SERIALIZER_CTX *ctx,
+                                        const unsigned char *kstr,
+                                        size_t klen);
+ int OSSL_SERIALIZER_CTX_set_passphrase_cb(OSSL_SERIALIZER_CTX *ctx, int enc,
+                                           pem_password_cb *cb, void *cbarg);
+ int OSSL_SERIALIZER_CTX_set_passphrase_ui(OSSL_SERIALIZER_CTX *ctx,
+                                           const UI_METHOD *ui_method,
+                                           void *ui_data);
+
+ #define OSSL_SERIALIZER_PUBKEY_TO_PEM_PQ "format=pem,type=public"
+ #define OSSL_SERIALIZER_PrivateKey_TO_PEM_PQ "format=pem,type=private"
+ #define OSSL_SERIALIZER_Parameters_TO_PEM_PQ "format=pem,type=domainparams"
+
+ #define OSSL_SERIALIZER_PUBKEY_TO_TEXT_PQ "format=text,type=public"
+ #define OSSL_SERIALIZER_PrivateKey_TO_TEXT_PQ "format=text,type=private"
+ #define OSSL_SERIALIZER_Parameters_TO_TEXT_PQ "format=text,type=domainparams"
+
+=head1 DESCRIPTION
+
+OSSL_SERIALIZER_CTX_new_by_EVP_PKEY() creates a B<OSSL_SERIALIZER_CTX>
+with a suitable attached output routine for B<EVP_PKEY>s.  It will
+search for a serializer implementation that matches the algorithm of
+the B<EVP_PKEY> and the property query given with I<propquery>.  It
+will prefer to find a serializer from the same provider as the key
+data of the B<EVP_PKEY> itself, but failing that, it will choose the
+first serializer that supplies a generic serializing function.
+
+If no suitable serializer was found, OSSL_SERIALIZER_CTX_new_by_EVP_PKEY()
+still creates a B<OSSL_SERIALIZER_CTX>, but with no associated
+serializer (L<OSSL_SERIALIZER_CTX_get_serializer(3)> returns NULL).
+This helps the caller distinguish between  an error when creating
+the B<OSSL_SERIALIZER_CTX>, and the lack the serializer support and
+act accordingly.
+
+OSSL_SERIALIZER_CTX_set_cipher_name() tells the implementation what
+cipher should be used to encrypt serialized keys.  If no cipher is
+specified, the implementation may still choose to serialize the key,
+but with unencrypted result.
```

Comment:
I need to have a think on this, and am currently pressed for time, so I'll get back

Reply:
The current consequence is sloppy...  `set_ctx_params` will return an error, but will do nothing to prevent serialization (which will become unencrypted)...  so OK, I should probably add an indicator that the caller tried to set a cipher, so serialization will also fail due to failed settings.

Reply:
I won't change this here, as we have so far *assumed* that whatever cipher / digest name we pass down to providers, they are going to fetch using libcrypto functions.  To modify this, we need a higher level discussion, which makes it out of scope for this PR.

Reply:
> I won't change this here, as we have so far assumed that whatever cipher / digest name we pass down to providers, they are going to fetch using libcrypto functions.

Have we??? I've never made that assumption!

Reply:
You have obviously not looked at the MAC stuff.  We pass down properties and engine names along with the cipher / digest name.  That's just as much an assumption as I'm making here.

Reply:
[rewriting]

Gah! I think I have mixed up two discussions here...

So that indicator I talked about, it's going to happen...

Reply:
I've made our serializers a little smarter.  It's possible to pass a NULL, which will clear away any cipher setting, and either way, the provider side context now registers a *intent* to encrypt, and refuses to do anything if the cipher isn't present even though the intent was to encrypt.  That should take care of an invlid cipher.